File: cpp/526_BeautifulArrangement.cpp
class Solution {
public:
    int count = 0;
    int countArrangement(int n) {
        bool used[n + 1];
        memset(used, false, n + 1);
        return backtrack(n, 1, used);
    }
    
    int backtrack(int n, int idx, bool used[]) {
        if (idx > n) return ++count;
        for (int i = 1; i <= n; ++i) {
            if (!used[i] && (idx % i == 0 || i % idx == 0)) {
                used[i] = true;
                backtrack(n, idx + 1, used);
                used[i] = false;
            }
        }
        return count;
    }
};

File: swift/526_BeautifulArrangement.swift
class Solution {
    func countArrangement(_ n: Int) -> Int {
        func backtrack(_ idx: Int) -> Int {
            if idx > n {
                count += 1
                return count
            }
            
            for i in 1...n {
                if !visited[i] && (i % idx == 0 || idx % i == 0) {
                    visited[i] = true
                    backtrack(idx + 1)
                    visited[i] = false
                }
            }
            return count
        }
        var count = 0
        var visited = [Bool](repeating: false, count: n + 1)
        return backtrack(1)
    }
}
